pocld/daemon.cc

```diff
@@ -415,6 +421,137 @@ int PoclDaemon::launch(std::string ListenAddress, struct ServerPorts &Ports) {
   return 0;
 }
 
+#ifdef ENABLE_VSOCK
+static int convertToVsockAddress(const std::string &VsockContextId, int Port,
+                                 struct sockaddr *Sockaddr,
+                                 socklen_t *SockaddrLen) {
+  if (Sockaddr == nullptr || SockaddrLen == nullptr)
+    return -1;
+
+  char *End = nullptr;
+  int Sibling = 0;
+  long CidValue;
+  struct sockaddr_vm *svm = (struct sockaddr_vm *)Sockaddr;
+
+  if (VsockContextId.empty()) {
+    *SockaddrLen = sizeof(*svm);
+    memset(svm, 0, *SockaddrLen);
+    svm->svm_family = AF_VSOCK;
+    svm->svm_cid = VMADDR_CID_ANY;
+    svm->svm_port = Port;
+
+    return 0;
+  }
+
+  /*
+   * Connection to a sibling VM
+   * VsockContextId is of the form "s<cid>"
+   */
+  if (VsockContextId[0] == 's') {
+#if LINUX_VERSION_CODE > KERNEL_VERSION(5, 11, 0)
+    Sibling = 1;
+#endif
+  }
+
+  CidValue = Sibling ? strtol(VsockContextId.c_str() + 1, &End, 10)
+                     : strtol(VsockContextId.c_str(), &End, 10);
+  if (VsockContextId.c_str() != End && *End == '\0' && CidValue <= UINT_MAX) {
+    *SockaddrLen = sizeof(*svm);
+    memset(svm, 0, *SockaddrLen);
+    svm->svm_family = AF_VSOCK;
+    svm->svm_cid = CidValue;
+    svm->svm_port = Port;
+#if LINUX_VERSION_CODE > KERNEL_VERSION(5, 11, 0)
+    if (Sibling) {
+      svm->svm_flags = VMADDR_FLAG_TO_HOST;
+    }
+#endif
+
+    return 0;
+  }
+  return -1;
+}
+
+int PoclDaemon::launchVsock(std::string ListenAddress,
```

Comment:
About two or three weeks ago, the launch function had the following parameters:
```
int PoclDaemon::launch(struct sockaddr_storage &base_addr,
                      socklen_t base_addrlen,
                      struct ServerPorts &ports)
```
I have a launch implementation that supports vsock, please refer to the following link:

[https://github.com/xvim/pocl/blob/vsock-lagacy/pocld/daemon.cc]

After the launch parameters were changed, it is not so easy to use only one launch. I am not sure if you accept the code similar above?

Reply:
Right, I changed that while refactoring so that the daemon could handle DNS names such as "localhost" resolving to multiple different addresses (e.g. IPv4 and IPv6) simultaneously. The way it's meant to work is that it simply takes all results and tries to listen on each produced sockaddr separately. If at least one of the resolved addresses can be successfully listened on that is considered a success.

What I care about is keeping this functionality. I don't have any attachment to the exact function signature of launch so long as this remains possible, hence why I suggested looking into `std::variant` or a tagged union so that it can be passed around as is and resolved to one or more results in the shared resolver.